lightning/src/ln/msgs.rs

```diff
@@ -52,49 +53,121 @@ pub enum DecodeError {
+/// available.
+///
+/// This Context represents when the Feature appears in the node_announcement message, as it is
+/// rumored around the P2P network.
+pub struct FeatureContextNode {}
+/// The context in which a Feature object appears determines which bits of features the node
+/// supports will be set. We use this when creating our own Feature objects to select which bits to
+/// set and when passing around Feature objects to ensure the bits we're checking for are
+/// available.
+///
+/// This Context represents when the Feature appears in the ChannelAnnouncement message, as it is
+/// rumored around the P2P network.
+pub struct FeatureContextChannel {}
+/// The context in which a Feature object appears determines which bits of features the node
+/// supports will be set. We use this when creating our own Feature objects to select which bits to
+/// set and when passing around Feature objects to ensure the bits we're checking for are
+/// available.
+///
+/// This Context represents when the Feature appears in an invoice, used to determine the different
+/// options available for routing a payment.
+///
+/// Note that this is currently unused as invoices come to us via a different crate and are not
+/// native to rust-lightning directly.
+pub struct FeatureContextInvoice {}
+
+/// An internal trait capturing the various future context types
+pub trait FeatureContext {}
+impl FeatureContext for FeatureContextInit {}
+impl FeatureContext for FeatureContextNode {}
+impl FeatureContext for FeatureContextChannel {}
+impl FeatureContext for FeatureContextInvoice {}
+
+/// An internal trait capturing FeatureContextInit and FeatureContextNode
+pub trait FeatureContextInitNode : FeatureContext {}
+impl FeatureContextInitNode for FeatureContextInit {}
+impl FeatureContextInitNode for FeatureContextNode {}
+
+/// Tracks the set of features which a node implements, templated by the context in which it
+/// appears.
+pub struct Features<T: FeatureContext> {
+	#[cfg(not(test))]
 	flags: Vec<u8>,
+	// Used to test encoding of diverse msgs
+	#[cfg(test)]
+	pub flags: Vec<u8>,
+	mark: PhantomData<T>,
+}
+
+impl<T: FeatureContext> Clone for Features<T> {
+	fn clone(&self) -> Self {
+		Self {
+			flags: self.flags.clone(),
+			mark: PhantomData,
+		}
+	}
+}
+impl<T: FeatureContext> PartialEq for Features<T> {
+	fn eq(&self, o: &Self) -> bool {
+		self.flags.eq(&o.flags)
+	}
+}
+impl<T: FeatureContext> fmt::Debug for Features<T> {
+	fn fmt(&self, fmt: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+		self.flags.fmt(fmt)
+	}
 }
 
-impl LocalFeatures {
-	/// Create a blank LocalFeatures flags (visibility extended for fuzz tests)
+impl<T: FeatureContextInitNode> Features<T> {
+	/// Create a blank Features flags (visibility extended for fuzz tests)
 	#[cfg(not(feature = "fuzztarget"))]
-	pub(crate) fn new() -> LocalFeatures {
-		LocalFeatures {
+	pub(crate) fn new() -> Features<T> {
+		Features {
 			flags: vec![2 | 1 << 5],
+			mark: PhantomData,
 		}
 	}
```

Comment:
The call sites of each `new` could be less verbose if these were module-level functions (e.g., `new_init_features()`). The return type could then be the exact specialization desired for the method.

Reply:
The callsites are much nicer with the type aliases, so I think this is resolved.